src/Features/Core/Portable/FullyQualify/AbstractFullyQualifyService.cs

```diff
@@ -0,0 +1,313 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeCleanup;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.Shared.Utilities.EditorBrowsableHelpers;
+
+namespace Microsoft.CodeAnalysis.CodeFixes.FullyQualify;
+
+internal abstract partial class AbstractFullyQualifyService<TSimpleNameSyntax> : IFullyQualifyService
+    where TSimpleNameSyntax : SyntaxNode
+{
+    private const int MaxResults = 3;
+
+    private const int NamespaceWithNoErrorsWeight = 0;
+    private const int TypeWeight = 1;
+    private const int NamespaceWithErrorsWeight = 2;
+
+    protected abstract bool CanFullyQualify(SyntaxNode node, [NotNullWhen(true)] out TSimpleNameSyntax? simpleName);
+    protected abstract Task<SyntaxNode> ReplaceNodeAsync(TSimpleNameSyntax simpleName, string containerName, bool resultingSymbolIsType, CancellationToken cancellationToken);
+
+    public async Task<FullyQualifyFixData?> GetFixDataAsync(
+        Document document, TextSpan span, bool hideAdvancedMembers, CancellationToken cancellationToken)
+    {
+        var client = await RemoteHostClient.TryGetClientAsync(document.Project, cancellationToken).ConfigureAwait(false);
+        if (client != null)
+        {
+            var result = await client.TryInvokeAsync<IRemoteFullyQualifyService, FullyQualifyFixData?>(
+                document.Project,
+                (service, solutionChecksum, cancellationToken) => service.GetFixDataAsync(solutionChecksum, document.Id, span, hideAdvancedMembers, cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!result.HasValue)
+                return null;
+
+            return result.Value;
+        }
+
+        return await GetFixDataInCurrentProcessAsync(document, span, hideAdvancedMembers, cancellationToken).ConfigureAwait(false);
+    }
```

Comment:
this is hte only new logic (remoting to OOP if possible to compute and return all the results).